README.md
# NAME

<div>
    <a href="https://travis-ci.org/PACKED-vzw/Catmandu-Store-RKD"><img src="https://travis-ci.org/PACKED-vzw/Catmandu-Store-RKD.svg?branch=master"></a>
</div>

Catmandu::RKD - Retrieve items from the RKD (Rijksbureau voor Kunsthistorische Documentatie) artist database

# SYNOPSIS

    # From the command line
    $ catmandu export VIAF --id 38885 to YAML
    ---
    - artist_link: https://rkd.nl/opensearch-eac-cpf?q=kunstenaarsnummer:38885
      description: hofschilder, schilder, tekenaar
      guid: https://rkd.nl/explore/artists/38885
      title: Hoey, Jan de
    ...

    # From a Catmandu Fix
    lookup_in_store(
        objectName,    # objectName is a field containing the RKD identifier
        RKD
    )

    # Retrieve a RKD record based on an artist name.
    rkd_name(path)     # path points to name

# MODULES

- [Catmandu::Store::RKD](https://metacpan.org/pod/Catmandu::Store::RKD)
- [Catmandu::Fix::rkd\_name](https://metacpan.org/pod/Catmandu::Fix::rkd_name)
- [Catmandu::RKD::API::Name](https://metacpan.org/pod/Catmandu::RKD::API::Name)

# AUTHORS

Pieter De Praetere, `pieter at packed.be`

# CONTRIBUTORS

Pieter De Praetere, `pieter at packed.be`

# COPYRIGHT AND LICENSE

This package is copyright (c) 2016 by PACKED vzw.
This is free software; you can redistribute it and/or modify it under the same terms as the Perl 5 programming language system itself.


Changes
Revision history for Catmandu-Store-RKD

{{$NEXT}}

0.04  2017-01-10 14:55:50 UTC
        - Rename package to Catmandu::RKD

0.03  2017-01-06 08:02:08 UTC
        - Updated dependency on LWP::Protocol::https

0.02  2017-01-05 14:30:56 UTC
        - Updated documentation

0.01  2017-01-05 14:24:36 UTC
        - Initial release


t/test.sh
catmandu convert JSON --fix test.fix to JSON < data.json

lib/Catmandu/Fix/rkd_name.pm
package Catmandu::Fix::rkd_name;

use strict;

use Catmandu::Sane;
use Moo;

use Catmandu::Fix::Has;
use Catmandu::Fix::Datahub::Util qw(declare_source);

with 'Catmandu::Fix::Base';

has path     => (fix_arg => 1);

sub emit {
    my ($self, $fixer) = @_;
    my $perl = '';

    $perl .= 'use Catmandu::Store::RKD::API::Name;';

    my $name = $fixer->generate_var();
    my $rkd = $fixer->generate_var();

    $perl .= "my ${name};";
    $perl .= declare_source($fixer, $self->path, $name);

    $perl .= "my ${rkd} = Catmandu::Store::RKD::API::Name->new(name_to_search => ${name});";

    $perl .= $fixer->emit_create_path(
        $fixer->var,
        $fixer->split_path($self->path),
        sub {
            my $root = shift;
            my $code = '';
            $code .= "${root} = ${rkd}->results;";

            return $code;
        }
    );

    return $perl;
}

1;

lib/Catmandu/Store/RKD/Bag.pm
package Catmandu::Store::RKD::Bag;

use Moo;
use LWP::UserAgent;

use Catmandu::RKD::API::Number;

use Catmandu::Sane;

with 'Catmandu::Bag';

sub generator {
    my $self = shift;
    Catmandu::NotImplemented->throw(
        message => 'Iterating over items in the store not supported.'
    );
}

sub get {
    my ($self, $id) = @_;
    my $sr = Catmandu::RKD::API::Number->new(
        artist_id => $id
    );
    return $sr->results;
}

sub add {
    my ($self, $data) = @_;
    Catmandu::NotImplemented->throw(
        message => 'Adding item to store not supported.'
    );
}

sub update {
    my ($self, $id, $data) = @_;
    Catmandu::NotImplemented->throw(
        message => 'Updating item in store not supported.'
    );
}

sub delete {
    my ($self, $id) = @_;
    Catmandu::NotImplemented->throw(
        message => 'Deleting item from store not supported.'
    );
}

sub delete_all {
    my $self = shift;
    Catmandu::NotImplemented->throw(
        message => 'Deleting items from store not supported.'
    );
}

1;


lib/Catmandu/Store/RKD/API/Name.pm
package Catmandu::Store::RKD::API::Name;

use Moo;
use LWP::UserAgent;
use XML::Struct qw(readXML);

use Catmandu::Sane;

use Catmandu::Store::RKD::API::Parse;
use Catmandu::Store::RKD::API::Extract;

has name_to_search => (is => 'ro', required => 1);

has engine   => (is => 'lazy');
has results  => (is => 'lazy');

sub _build_engine {
    my $self = shift;
    my $engine = LWP::UserAgent->new();
    return $engine;
}

sub _build_results {
    my $self = shift;
    my $items_raw = $self->__search();
    my $items = [];
    
    # The results are in $items_raw->{'items'}
    my $parser = Catmandu::Store::RKD::API::Parse->new(results => $items_raw);
    my $parsed_items = $parser->items;
    push @{$items}, @{$parsed_items->{'items'}};

    # The API paginates, but we want all the results. So continue as long as there are items
    while ($parsed_items->{'total'} > $parsed_items->{'start'} * $parsed_items->{'per_page'}) {

        my $startIndex = $parsed_items->{'start'} + $parsed_items->{'per_page'};
        $items_raw = $self->__search($startIndex);

        $parser = Catmandu::Store::RKD::API::Parse->new(results => $items_raw);
        $parsed_items = $parser->items;

        push @{$items}, @{$parsed_items->{'items'}};
    }

    # Return [{
    #    guid        => foo,
    #    title       => bar,
    #    description => biz,
    #    artist_link => xyz
    #}]
    my $extracter = Catmandu::Store::RKD::API::Extract->new(results => $items);
    return $extracter->items;
}

##
# Query the API
sub __search {
    my ($self, $startIndex) = @_;
    my $url;
    if (defined($startIndex)) {
        my $template = 'http://opendata.rkd.nl/opensearch/artists/eac-cpf?q=naamdeel:(%s)&startIndex=%s';
        $url = sprintf($template, $self->name_to_search, $startIndex);
    } else {
        my $template = 'http://opendata.rkd.nl/opensearch/artists/eac-cpf?q=naamdeel:(%s)';
        $url = sprintf($template, $self->name_to_search);
    }
    my $response = $self->engine->get($url);
    if ($response->is_success) {
        return $response->decoded_content;
    } else {
        Catmandu::HTTPError->throw({
                code             => $response->code,
                message          => $response->status_line,
                url              => $response->request->uri,
                method           => $response->request->method,
                request_headers  => [],
                request_body     => $response->request->decoded_content,
                response_headers => [],
                response_body    => $response->decoded_content,
            });
        return undef;
    }
}

1;

lib/Catmandu/Store/RKD.pm
package Catmandu::Store::RKD;

our $VERSION = '0.04';

use Catmandu::Sane;

use Moo;
use Catmandu::Store::RKD::Bag;

with 'Catmandu::Store';

1;

__END__

=head1 NAME

=for html <a href="https://travis-ci.org/PACKED-vzw/Catmandu-Store-RKD"><img src="https://travis-ci.org/PACKED-vzw/Catmandu-Store-RKD.svg?branch=master"></a>

Catmandu::Store::RKD - Retrieve items from the RKD (Rijksbureau voor Kunsthistorische Documentatie) artist database

=head1 SYNOPSIS

    # From the command line
    $ catmandu export VIAF --id 38885 to YAML
    ---
    - artist_link: https://rkd.nl/opensearch-eac-cpf?q=kunstenaarsnummer:38885
      description: hofschilder, schilder, tekenaar
      guid: https://rkd.nl/explore/artists/38885
      title: Hoey, Jan de
    ...

    # From a Catmandu Fix
    lookup_in_store(
        objectName,    # objectName is a field containing the RKD identifier
        RKD
    )

    # From Perl code
    use Catmandu;

    my $store = Catmandu->store('RKD')->bag;

    my $item = $store->get('38885');

    print $item->[0]->{'title'} , "\n";  # Hoey, Jan de

=head1 DESCRIPTION

A Catmandu::Store::VIAF is a Perl package that can query the
L<RKD|https://rkd.nl/nl/collecties/overige-databases/open-search-rkdartists> artist
database by I<kunstenaarsnummer>.

This store supports only one method C<get> to retrieve an AAT record by its identifier

For some names, it can/will return multiple possibilities. You must determine yourself which one is the 'correct' one.

=head1 CONFIGURATION

No options.

=head1 METHODS

=head2 new(%configuration)

Create a new Catmandu::Store::RKD

=head2 get($id)

The store takes an artist id (I<kunstenaarsnummer>) and performs a lookup to the RKD artists database. It
returns an array containing either one or no results.  Every result is of the form:

    {
        'title'       => 'Name of the person',
        'description' => 'Short description, as provided by RKD',
        'artist_link' => 'Link to the artist using the artist id',
        'guid'        => 'Permalink to the record'
    }

=head2 add()

Not supported

=head2 delete()

Not supported

=head2 each()

Not supported

=head1 SEE ALSO

L<Catmandu>
L<Catmandu::Fix::rkd_name>

=head1 AUTHORS

Pieter De Praetere, C<< pieter at packed.be >>

=head1 CONTRIBUTORS

Pieter De Praetere, C<< pieter at packed.be >>

=head1 COPYRIGHT AND LICENSE

This package is copyright (c) 2016 by PACKED vzw.
This is free software; you can redistribute it and/or modify it under the same terms as the Perl 5 programming language system itself.

=cut
